### PR TITLE
Fix a crash bug in MXNet due to early free of tensor object

### PR DESCRIPTION
**Pull request by apeforest**
This PR includes the following changes:
1) Avoid early destruction of NDArray invoked by Python garbage collection by making a copy using shared_ptr
2) Simplify code by removing unnecessary template class
3) Code refactoring
4) Fixed a bug in ResizeND mentioned in https://github.com/horovod/horovod/issues/1409

This PR fixes https://github.com/horovod/horovod/issues/1533

@eric-haibin-lin @karan6181 @yuxihu Please help review!

### CODE REVIEW

File: horovod/mxnet/adapter.cc [hunk 1]
```diff
@@ -53,13 +53,13 @@ MXPersistentBuffer::AccessData(std::shared_ptr<OpContext> context) const {
   return buffer_;
 }
 
-template <class T> MXTensor<T>::MXTensor(T* tensor) : tensor_(tensor) {}
+MXTensor::MXTensor(NDArray* tensor) : tensor_(tensor) {}
```
Comment (karan6181) [inline horovod/mxnet/adapter.cc]: What happens if it's MXNet numpy array ?

File: horovod/mxnet/adapter.cc [hunk 2]
```diff
@@ -70,76 +70,62 @@ template <class T> const TensorShape MXTensor<T>::shape() const {
...
-MXOpContext<T>::MXOpContext(int device, T* output)
+MXOpContext::MXOpContext(int device, NDArray* output)
     : device_(device), output_(output) {}
 
-template <class T>
-Status
-MXOpContext<T>::AllocatePersistent(int64_t size,
-                                   std::shared_ptr<PersistentBuffer>* tensor) {
+Status MXOpContext::AllocatePersistent(int64_t size,
+  std::shared_ptr<PersistentBuffer>* tensor) {
```
Comment (karan6181) [inline horovod/mxnet/adapter.cc]: Nit: Spacing

File: horovod/mxnet/mpi_ops.cc [hunk 3]
```diff
@@ -69,26 +69,27 @@ void DoHorovodOperation(void*, void* on_complete_ptr, void* param) {
...
-      hvd_output = std::make_shared<MXTensor<NDArray>>(output);
+      hvd_output = std::make_shared<MXTensor>(output);
       enqueue_result = EnqueueTensorAllreduce(
           hvd_context, hvd_tensor, hvd_output, nullptr, name, device,
           [on_complete](const Status& status) {
             InvokeCompleteCallback(on_complete, status);
       });
       break;
     case OperationType::ALLGATHER:
+      hvd_output = std::make_shared<MXTensor>(output);
```
Comment (karan6181) [inline horovod/mxnet/mpi_ops.cc]: No usage of `hvd_output` in `case OperationType::ALLGATHER` ?

File: horovod/mxnet/adapter.cc (hunk 1 above)
Comment (apeforest) [reply horovod/mxnet/adapter.cc]: Is there a Numpy array class in MXNet? I think it still uses NDArray to store tensors.

**Comment by wangli1426** [comment]
May I ask when this PR may be merged? I really need this bugfix. :)

**Comment by apeforest** [comment]
@karan6181 @eric-haibin-lin gentle ping for review :)

**Comment by karan6181** [comment]
LGTM. Thank You for your contributions!

**Comment by tgaddair** [comment]
@apeforest please try rebasing off master.  Unit tests should be fixed following #1642.

**Pull request merged**